doc/reference/misc/timeutil.rst

```diff
@@ -0,0 +1,289 @@
+(or "local time") which is the time at a particular location. Standard
+time has a fixed offset from UT at any given instant, primarily
+influenced by longitude, but the offset may be adjusted ("daylight
+saving time") to align standard time to the local solar time. In a sense
+local time is "more discontinuous" than UT.
+
+`POSIX Time <timeutil-RFC8536s2>`_ is a time scale that counts seconds since
+the "POSIX epoch" at 1970-01-01T00:00:00Z (i.e. the start of 1970
+UTC). `UNIX Time <timeutil-RFC8536s2>`_ is an extension of POSIX time using
+negative values to represent times before the POSIX epoch. Both of these
+scales assume that every day has exactly 86400 seconds. In normal use
+instants in these scales correspond to times in the UTC scale, so they
+inherit the discontinuity.
+
+The continuous analogue is `UNIX Leap Time <timeutil-RFC8536s2>`_ which
+is UNIX time plus all leap-second corrections added after the POSIX
+epoch (when TAI-UTC was 8 s).
+
+Example of Time Scale Differences
+---------------------------------
+
+A positive leap second was introduced at the end of 2016, increasing the
+difference between TAI and UTC from 36 seconds to 37 seconds. There was
+no leap second introduced at the end of 1999, when the difference
+between TAI and UTC was only 32 seconds. The following table shows
+relevant civil and epoch times in several scales:
+
+==================== ========== =================== ======= ==============
+UTC Date             UNIX time  TAI Date            TAI-UTC UNIX Leap Time
+==================== ========== =================== ======= ==============
+1970-01-01T00:00:00Z 0          1970-01-01T00:00:08 +8      0
+1999-12-31T23:59:28Z 946684768  2000-01-01T00:00:00 +32     946684792
+1999-12-31T23:59:59Z 946684799  2000-01-01T00:00:31 +32     946684823
+2000-01-01T00:00:00Z 946684800  2000-01-01T00:00:32 +32     946684824
+2016-12-31T23:59:59Z 1483228799 2017-01-01T00:00:35 +36     1483228827
+2016-12-31T23:59:60Z undefined  2017-01-01T00:00:36 +36     1483228828
+2017-01-01T00:00:00Z 1483228800 2017-01-01T00:00:37 +37     1483228829
+==================== ========== =================== ======= ==============
+
+Functional Requirements
+-----------------------
+
+The Zephyr tick counter has no concept of leap seconds or standard time
+offsets and is a continuous time scale. However it can be relatively
+inaccurate, with drifts as much as three minutes per hour (assuming an
+RC timer with 5% tolerance).
+
+There are two stages required to support conversion between Zephyr time
+and common human time scales:
+
+* Translation between the continuous but inaccurate Zephyr time scale
+  and an accurate external stable time scale;
+* Translation between the stable time scale and the (possibly
+  discontinuous) civil time scale.
+
+The API around :c:func:`timeutil_sync_state_update()` supports the first
+step of converting between continuous time scales.
+
+The second step requires external information including schedules of
```

Comment:
So right now this API could still cause issues if the external time source wouldn't be monotonic (like most wallclock time sources, such as RTC, might not be)?

Reply:
Is there a way that adding such functionality would allow support of a discontinuous time source?

Reply:
I don't believe it belongs in this API.  You'd use this API to get a timestamp in a continuous scale like TAI, then do adjustments using external tables and rules.

E.g. if local time is in TAI or GPS time (continuous monotonic), you'd need the leap second tables referenced elsewhere to figure out what the UTC offset is for that time.  Then you'd use the UTC time in conjunction with tables/rules for the local time zone to get the corresponding local time.  You can reverse that, with the problem that some local times are either ambiguous (1:30 am during a "fall back" adjustment) or non-existent (1:30 am during "spring forward").

If you're using this API with a UTC time source, which is fairly likely, you'd still need the leap second tables to deal with the discontinuity when a leap second occurred.  You can see that in the table up around like 215.

Reply:
(Quick add-on: the RTC in STM32 SoCs can be non-monotonic, if the steps are taken to implement daylight saving time, and these instructions can also be memorized (by bit-setting). However, I believe right now that is not part of the configuration, so won't be a problem right now, if the RTC counter driver would be used)

I was asking the question because I was wondering if that would make it possible for us to still use the API, but since we discussed this above, that's solved. 

If you do end up doing the long-duration data collection runs after the API has been merged, please do share! (I guess it could also be useful for the potential switching clock source in Nordic, right? Since that would change the base instant and skew.)